File: Leetcode/java/Binary Tree Maximum Path Sum.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int maxPathSum(TreeNode root) {
        ans = -0x3f3f3f3f;
        do_maxPathSum(root);
        return ans;
    }
    
    private int do_maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int max_left  = Math.max(do_maxPathSum(root.left), 0);
        int max_right = Math.max(do_maxPathSum(root.right), 0);
        ans = Math.max(ans, max_left + root.val + max_right);
        return root.val + Math.max(max_left, max_right);
    }
    
    private int ans;
} 

File: LintCode/cpp/Binary Tree Maximum Path Sum.cc
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: An integer
     */
    int maxPathSum(TreeNode *root) {
        ans = numeric_limits<int>::min();
        do_maxPathSum(root);
        return ans;
    }
private:
    int do_maxPathSum(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        int left = max(0,
            do_maxPathSum(root->left));
        int right = max(0,
            do_maxPathSum(root->right));
        ans = max(ans, root->val + left + right);
        return root->val + max(left, right);
    }
private:
    int ans;
};
